pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,359 @@
+// 	servicesGVR               = corev1.SchemeGroupVersion.WithResource("services")
+// 	persistentVolumeClaimsGVR = corev1.SchemeGroupVersion.WithResource("persistentvolumeclaims")
+// )
+//
+// type hackyAdmissionAttributesForBuiltInTypes struct {
+// 	admission.Attributes
+// 	builtInCurrent runtime.Object
+// 	builtInOld     runtime.Object
+// }
+//
+// func (h *hackyAdmissionAttributesForBuiltInTypes) GetObject() runtime.Object {
+// 	if h.builtInCurrent != nil {
+// 		return h.builtInCurrent
+// 	}
+//
+// 	obj, converted := objectFor(h.GetResource(), h.Attributes.GetObject())
+// 	if converted {
+// 		h.builtInCurrent = obj
+// 	}
+//
+// 	return obj
+// }
+//
+// func (h *hackyAdmissionAttributesForBuiltInTypes) GetOldObject() runtime.Object {
+// 	if h.builtInOld != nil {
+// 		return h.builtInOld
+// 	}
+//
+// 	obj, converted := objectFor(h.GetResource(), h.Attributes.GetOldObject())
+// 	if converted {
+// 		h.builtInOld = obj
+// 	}
+//
+// 	return obj
+// }
+//
+// func objectFor(gvr schema.GroupVersionResource, obj runtime.Object) (out runtime.Object, converted bool) {
+// 	switch gvr {
+// 	case podsGVR:
+// 		pod := &corev1.Pod{}
+// 		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(runtime.Unstructured).UnstructuredContent(), pod); err != nil {
+// 			utilruntime.HandleError(fmt.Errorf("error converting unstructured to pod: %w", err))
+// 		}
+//
+// 		return pod, true
+// 	case servicesGVR:
+// 		service := &corev1.Service{}
+// 		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(runtime.Unstructured).UnstructuredContent(), service); err != nil {
+// 			utilruntime.HandleError(fmt.Errorf("error converting unstructured to service: %w", err))
+// 		}
+//
+// 		return service, true
+// 	case persistentVolumeClaimsGVR:
+// 		pvc := &corev1.PersistentVolumeClaim{}
+// 		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(runtime.Unstructured).UnstructuredContent(), pvc); err != nil {
+// 			utilruntime.HandleError(fmt.Errorf("error converting unstructured to persistent volume claim: %w", err))
+// 		}
+//
+// 		return pvc, true
+// 	}
+//
+// 	return obj, false
+// }
+
+// getOrCreateDelegate creates a resourcequota.QuotaAdmission plugin for clusterName.
+func (k *KubeResourceQuota) getOrCreateDelegate(clusterName logicalcluster.Name) (*stoppableQuotaAdmission, error) {
+	var delegate *stoppableQuotaAdmission
+
+	k.lock.RLock()
+	delegate = k.delegates[clusterName]
+	k.lock.RUnlock()
+
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	delegate = k.delegates[clusterName]
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	// Set up a context that is cancelable and that is bounded by k.ServerDone
+	ctx, cancel := context.WithCancel(context.Background())
+	go func(done <-chan struct{}) {
+		<-done
+		cancel()
+	}(k.serverDone)
+
+	quotaAdmission, err := resourcequota.NewResourceQuota(k.userSuppliedConfiguration, 5, ctx.Done())
```

Comment:
Is that 5 workers per?

Reply:
our workspaces are smaller than kube clusters. Can we reduce 5 ?

Reply:
Have lowered to 1 for now (in next push)